docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,532 @@
+    ```go
+    // Current Code
+    // sdk.go.tpl    
+    func (rm *resourceManager) sdkCreate(
+	ctx context.Context,
+	desired *resource,
+    ) (created *resource, err error) {
+        ...
+        {{ GoCodeSetCreateOutput .CRD "resp" "ko" 1 false }}
+        ...
+    }
+    ```
+
+    ```go
+    // Proposed Code
+    // sdk.go.tpl
+    func (rm *resourceManager) sdkCreate(
+	ctx context.Context,
+	desired *resource,
+    ) (created *resource, err error) {
+        ...
+        // NOTE: The performSpecUpdate boolean param will be removed
+        {{ GoCodeSetCreateOutput .CRD "resp" "ko" 1 }}
+        ...
+    }
+    ```
+
+2. **Provide a new "LateInitializationConfig" struct inside generator config for AWS service teams to provide which spec fields will be late initialized.** These fields will be the ones which are late initialized by the server but not present in the Create and Update response.
+
+    These fields will be populated by making the readOne call. AWS service teams can also provide a delay after which these fields will be late initialized.
+    LateInitializationConfig will be added as part of FieldConfig. 
+
+    ```go
+        // LateInitializeConfig contains instructions for how to handle the
+        // retrieval and setting of server-side defaulted fields.
+        type LateInitializationConfig struct {
+            // DelaySeconds provides the number of seconds to wait to set
+            // late initialized fields after successful Create/Update call
+            DelaySeconds int `json:"delay_seconds,omitempty"`
+        }
+
+        type FieldConfig struct {
+            LateInitialize *LateInitializeConfig `json:"late_initialize,omitempty"`
+        }
+    ```
+
+3. **Provide 2 new hooks to AWS service teams to customize the code generation in step 4**
+    
+    These 2 hooks will be:
+
+    a) **late_initialize_pre_read_one**:- This hook will provide the custom code to be inserted **before** setting server side defaults for missing optional spec fields.
+
+    b) **late_initialize_post_read_one**: This hook will provide the custom code to be inserted **after** setting server side defaults for missing optional spec fields.
+
+4. Add `AWSResourceManager.LateInitialize()` method to handle server side defaults for fields mentioned in LateInitializationConfig.
+
+* Proposed Pseudo Code
+    ```go
+    // manager.go.tpl
+    func (rm *resourceManager) LateInitialize(
+	ctx context.Context,
+    latest acktypes.AWSResource,
+    ) (acktypes.AWSResource, error) {
+        {{- if $hookCode := Hook .CRD "late_initialize_pre_read_one" }}
+            {{ $hookCode }}
+        {{- end }}
+
+        // 1. Filter all the fields which have LateInitializationConfig.
+        // 2. Filter out the fields from step1 with nil value in latest object.
```

Comment:
Filter out the fields with `non-nil` value? We *want* the fields that have `nil` value?

Reply:
Yeah, which shows their absence and eligibility to be late-initialized.

Reply:
Sorry, by this I meant it currently reads `filter out ... with nil value` whereas I think you mean `filter out ... with non-nil value`. "Filter out" meaning to "discard those with"